theano/sandbox/cuda/cula.py

```diff
@@ -140,3 +143,142 @@ def cula_gpu_solve(A_, b_, trans='T'):
         return thunk
 
 gpu_solve = GpuSolve()
+
+
+class GpuCholesky(GpuOp):
+    """
+    CULA GPU Choleksy factorisation op.
+
+    Given a real positive definite matrix `A` returns either a lower
+    triangular matrix `L` such that `A == dot(L, L.T)` if `lower == True`
+    else returns an upper triangular matrix `U` such that `A == dot(U.T, U)`
+    if `lower == False`.
+
+    Parameters
+    ----------
+    lower
+        Whether to return a lower rather than upper triangular decomposition.
+
+    """
+
+    __props__ = ('lower',)
+
+    def __init__(self, lower=True):
+        self.lower = lower
+        super(GpuCholesky, self).__init__()
+
+    def output_type(self, inp):
+        return CudaNdarrayType(broadcastable=[False] * inp.type.ndim)
+
+    def make_node(self, inp):
+        inp = as_cuda_ndarray_variable(inp)
+
+        assert inp.ndim == 2
+        return theano.Apply(self, [inp], [self.output_type(inp)()])
+
+    def make_thunk(self,
+                   node,
+                   storage_map, _,
+                   no_recycling=[]):
+
+        # Initialize CULA the first time it is needed
+        global cula_initialized
+
+        if not cula_available:
+            raise RuntimeError('Cula is not available and '
+                               'GpuCholesky Op can not be constructed.')
+
+        if not pycuda_available:
+            raise RuntimeError('pycuda is not available and '
+                               'GpuCholesky Op can not be constructed.')
+
+        if not cula_initialized:
+            cula.culaInitialize()
+            cula_initialized = True
+
+        # import util function here to avoid circular import errors
+        from theano.misc.pycuda_utils import to_gpuarray
+
+        inputs = [storage_map[v] for v in node.inputs]
+        outputs = [storage_map[v] for v in node.outputs]
+
+        def thunk():
+
+            # Matrix to decompose is first (and only) input
+            A = inputs[0][0]
+
+            # Cholesky decomposition of matrix is first (and only) output
+            A_chol = outputs[0]
+
+            # CULA Cholesky function is destructive as it assigns calculated
+            # decomposition to lower / upper triangle of input array therefore
+            # force copy of array with allocation of a new C-contiguous buffer.
+            # CULA operation assumes Fortran ordering however input matrix must
+            # be symmetric therefore no need to alter order.
```

Comment:
Why inputs must be symmetric? I don't think this is right. I could easily add a parameter order that default to C, but could be F as numpy. This would make the copy a fortran order version.